source/extensions/filters/network/redis_proxy/conn_pool_impl.cc

```diff
@@ -122,6 +143,77 @@ InstanceImpl::ThreadLocalPool::makeRequest(const std::string& key,
     client->redis_client_->addConnectionCallbacks(*client);
   }
 
+  // cache host for quick lookup via address for redirection support
+  if (host->address()) {
+    auto hostCachedByAddress = host_address_map_.find(host->address()->asString());
+    if (hostCachedByAddress == host_address_map_.end()) {
+      host_address_map_[host->address()->asString()] = host;
+    }
+  }
+
+  return client->redis_client_->makeRequest(request, callbacks);
+}
+
+Common::Redis::Client::PoolRequest*
+InstanceImpl::ThreadLocalPool::makeRequestToHost(const std::string& host_address,
+                                                 const Common::Redis::RespValue& request,
+                                                 Common::Redis::Client::PoolCallbacks& callbacks) {
+  if (cluster_ == nullptr) {
+    ASSERT(client_map_.empty());
+    ASSERT(host_set_member_update_cb_handle_ == nullptr);
+    return nullptr;
+  }
+
+  auto colon_pos = host_address.rfind(":");
+  if ((colon_pos == std::string::npos) || (colon_pos == (host_address.size() - 1))) {
+    return nullptr;
+  }
+
+  std::string ip_address = host_address.substr(0, colon_pos);
+  bool ipv6 = (ip_address.find(":") != std::string::npos);
+  std::string host_address_map_key;
+
+  if (!ipv6) {
+    host_address_map_key = host_address;
+  } else {
+    // find our canonical IPv6 address string representation by
+    // creating the address from the Redis server info...
+    std::string ip_port = host_address.substr(colon_pos + 1);
+    try {
+      Network::Address::Ipv6Instance ipv6_addr(ip_address, std::stoi(ip_port));
+      host_address_map_key = ipv6_addr.asString();
+    } catch (EnvoyException&) {
+      // bad address
+      return nullptr;
+    }
+  }
+
+  auto it = host_address_map_.find(host_address_map_key);
+  if (it == host_address_map_.end()) {
+    // not known to the cluster manager, create a new host and insert it into the map
+    try {
+      Upstream::HostSharedPtr new_host{new Upstream::HostImpl(
+          cluster_->info(), "", Network::Utility::resolveUrl("tcp://" + host_address_map_key),
```

Comment:
I think it's probably easier to just create the ipv4 and ipv6 addresses instead of using the resolveUrl

Reply:
fine, but I think we should only take on the construction of an IPv4 address if we have to create a new host, while we need to create an IPv6 address just to look up the host in the host_address_map_ properly (with a canonical string representation) ... I think the resulting code will actually be clunkier, but I'm not wedded to this.